### PR TITLE
make ObjectReference field ownership granular

### PR DESCRIPTION
**Pull request by alexzielenski**
#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:

This change changes the `structType` of `ObjectReference` from `atomic` (atomic added in v1.22 #100684). 

Something this PR is particularly concerned with is the transition path from users of 1.22 with atomic ownership, to having granular field management setting. An integration test was added to ensure the expected behavior.

Without this fix, users of SSA will find since the object is atomic, unspecified fields are cleared by the apiserver when applying changes to objects. This is a problem because there are often multiple agents writing to an `ObjectReference`. For example, in the case of `PersistentVolume`'s `claimRef`, the `uid` field and others might be populated by `kube-controller-manager` but the `name` or `namespace` fields would be specified by a user.

#### Which issue(s) this PR fixes:

https://github.com/fluxcd/flux2/issues/2827
issues with this behavior reported in [this comment](https://github.com/kubernetes/kubernetes/pull/100684#discussion_r888465096)

#### Does this PR introduce a user-facing change?

```release-note
Changes ownership semantics of PersistentVolume's spec.claimRef from `atomic` to `granular`.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
Note for Server-Side Apply Users: Following this change, existing atomic owners of PersistentVolumeClaim `spec.claimRef` fields will not own any of the nested fields. Nested fields will have whatever owner may be already specified in managed fields, or whoever claims the field first.
```

/assign @apelisse

### CODE REVIEW

File: staging/src/k8s.io/api/core/v1/types.go [hunk 1]
```diff
@@ -5635,7 +5635,7 @@ type ServiceProxyOptions struct {
 // Instead of using this type, create a locally provided and used type that is well-focused on your reference.
 // For example, ServiceReferences for admission registration: https://github.com/kubernetes/api/blob/release-1.17/admissionregistration/v1/types.go#L533 .
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +structType=atomic
+// +structType=granular
```
Comment (apelisse) [inline staging/src/k8s.io/api/core/v1/types.go:5638]: Do we want to keep it called out or should we just remove the structType marker altogether?

**Comment by apelisse** [comment]
Thanks @alexzielenski, I'm curious what @liggitt will say :-)

**Comment by aojea** [comment]
Kmews*

File: staging/src/k8s.io/client-go/applyconfigurations/internal/internal.go [hunk 2]
```diff
@@ -5248,7 +5248,7 @@ var schemaYAML = typed.YAMLObject(`types:
     - name: uid
       type:
         scalar: string
-    elementRelationship: atomic
+    elementRelationship: separable
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/applyconfigurations/internal/internal.go:5251]: what are the implications for client-go consumers with old versions of this?

File: test/integration/apiserver/apply/apply_test.go [hunk 3]
```diff
@@ -3408,3 +3408,206 @@ func TestSubresourceField(t *testing.T) {
...
+	// if err != nil {
+	// 	t.Fatalf("Failed marshal yaml: %v", err)
+	// }
+
+	// Write our raw object in with its managed fields
+	originalObj, err := client.CoreV1().RESTClient().
+		Post().
+		Param("fieldManager", "apply_test").
+		Resource("persistentvolumes").
+		Body(oldPersistentVolume).
```
Comment (liggitt) [inline test/integration/apiserver/apply/apply_test.go:3579]: I'm surprised this actually worked, trying to create an object with an existing uid set. Did you verify the persisted/returned object had the managedFields set you expected?

File: test/integration/apiserver/apply/apply_test.go [hunk 4]
```diff
@@ -3408,3 +3408,206 @@ func TestSubresourceField(t *testing.T) {
...
+
+	// oldPersistentVolume, err := yamlutil.ToJSON(oldPersistentVolume)
+	// if err != nil {
+	// 	t.Fatalf("Failed marshal original yaml: %v", err)
+	// }
+
+	// originalPV, err = yamlutil.ToJSON(originalPV)
+	// if err != nil {
+	// 	t.Fatalf("Failed marshal yaml: %v", err)
+	// }
```
Comment (liggitt) [inline test/integration/apiserver/apply/apply_test.go:3572]: drop commented lines?

**Comment by liggitt** [comment]
I see the case for this being a bug in PV, but most of the other uses still seem like atomic is correct.

Can we drop the structType and keep the other uses atomic by decorating the field at point of use?

**Comment by apelisse** [comment]
> I see the case for this being a bug in PV, but most of the other uses still seem like atomic is correct.

Wouldn't you agree that we can't generally assume the UID  will be set by the same controller/actor as the gvk?

**Comment by apelisse** [comment]
> > Wouldn't you agree that we can't generally assume the UID will be set by the same controller/actor as the gvk?
> 
> Yeah, probably not generally... that's an argument in favor of dropping the structType decoration and adding the field-level decorations to indicate it is atomic in most of the places we use the struct (if that even works)

Yeah, that sounds reasonable. I want to note that non-atomic gives much more control than atomic, there's no easy way to fix a zealous atomic marker with server-side apply, if you want to set the whole non-atomic struct, you can still set uid to null or something to unset it.

File: test/integration/apiserver/apply/apply_test.go (hunk 3 above)
Comment (apelisse) [reply test/integration/apiserver/apply/apply_test.go:3579]: Yeah, I would expect all the fields set in the request to be owned by the POSTer here.

Comment (apelisse) [reply test/integration/apiserver/apply/apply_test.go:3579]: I think to be sure that you have the managedfields that you want, it'd be simpler to:
1. Create the object,
2. Send a patch (non-apply) that changes nothing in the object but the managedFields

**Comment by alexzielenski** [comment]
> > > I see the case for this being a bug in PV, but most of the other uses still seem like atomic is correct.
> > 
> > 
> > Wouldn't you agree that we can't generally assume the UID will be set by the same controller/actor as the gvk?
> 
> Yeah, probably not generally... that's an argument in favor of dropping the structType decoration and adding the field-level decorations to indicate it is atomic in most of the places we use the struct (if that even works)

I've attempted to add field-level decorators onto `PersistentVolume`'s `claimRef` which do have an effect on the resultant openapi; but they do not seem to override the type-level `atomic` specification.

Another alternative to making `ObjectReference` granular would be to make a `PersistentVolumeClaimReference` type for this use case? I am unsure how to assess which is more appropriate/less disruptive.

**Comment by apelisse** [comment]
Let's look at occurences of ObjectReference and maybe see how they can be used.

File: test/integration/apiserver/apply/apply_test.go (hunk 3 above)
Comment (alexzielenski) [reply test/integration/apiserver/apply/apply_test.go:3574]: Added separate update for managed fields. also added checks at the end to verify the managed fields were to be expected

**Comment by alexzielenski** [comment]
re: @liggitt's comments

> Can we drop the structType and keep the other uses atomic by decorating the field at point of use?

This is the current state of the PR. I have removed the `structType` annotation from `ObjectReference` and added `field-level` annotations at point of use. However, done this way, the field-level annotations are not respected. I'm not sure what the proper way to do this would be using the current tools.

> Alternatively, can we keep structType as atomic and override the PV use to be granular by decorating the field... because this is a core type, changing structType will have implications for all out-of-tree consumers that compile this into their APIs as well

I've attempted this and unfortunately the k8s apiserver does not give precedence to the field-level annotation on the PV's field over the `ObjectReference`'s type-level annotation.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 5]
```diff
@@ -337,6 +337,7 @@ type PersistentVolumeSpec struct {
 	// claim.VolumeName is the authoritative bind between PV and PVC.
 	// More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#binding
 	// +optional
+	// +structType=granular
```
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go:340]: I'm not sure our SSA code honors this annotation on a field... based on your experiments, it seems like it doesn't?

Comment (alexzielenski) [reply staging/src/k8s.io/api/core/v1/types.go:340]: Correct, SSA does not honor it. I only pushed the code to share the diff/experiement while we decide how to move forward. To me our choices seem to be to either:

1.) Drop `atomic` from ObjectReference, make SSA honor field-level annotations and attach `atomic` as a field-level reference where necessary.
2.) Give `PersistentVolume` its own `ObjectReference` type

Comment (liggitt) [reply staging/src/k8s.io/api/core/v1/types.go:340]: 1 seems best, though its unfortunate it requires SSA changes

2 and 3 leave ObjectReference in a non-ideal state for other out of tree consumers

3 is too disruptive for a patch backport (breaks go compatibility of API types)

Comment (alexzielenski) [reply staging/src/k8s.io/api/core/v1/types.go:340]: I am in favor of #1 in two parts:

a. Drop `atomic` from `ObjectReference`
b. Update `SSA` to allow field-level annotations of `structType`

Comment (apelisse) [reply staging/src/k8s.io/api/core/v1/types.go:340]: Agreed, we need to stop the bleeding with a., we can do cosmetic changes with b. as needed later.

Comment (liggitt) [reply staging/src/k8s.io/api/core/v1/types.go:340]: if we change a and backport it, we've also modified all the other API uses of this type... have we considered whether *those* changes will have negative consequences?

Comment (apelisse) [reply staging/src/k8s.io/api/core/v1/types.go:340]: I'm not aware of anything client-wise (especially in client-go) using the structType to change the generation of anything. I'm not aware of anything client-side that uses this specifically to make any determination outside possibly of calling the Update/Apply function of the FieldManager. That includes kubectl who never uses this to do anything relative to Strategic-Merge-Patch or anything else.

The FieldManager has some logic to deal with such changes in schema (we have a test in the smd repo for these, please we now have a test in this PR to verify that it works with actual types). The main consequence of this change is that we'll have a dangling ownership of the struct which isn't going to be honored. It'll probably stay "forever" without expected impact.

**Comment by alexzielenski** [comment]
PR has been updated to use new SSA feature for a field-level specification to override a type-level `structType`. Now `PersistentVolumeSpec.claimRef` uses a granular `ObjectReference`, despite `ObjectReference` being marked as `atomic`.

cc @liggitt @apelisse

**Comment by liggitt** [comment]
/lgtm
/approve

thanks for going the extra mile for compatibility here, and for the test coverage

**Comment by liggitt** [comment]
/milestone v1.25

adding to the milestone, since this is a bugfix for a regression in behavior that we're likely to try to backport

**Comment by janetkuo** [comment]
Is this ever backported to older versions? @alexzielenski

**Comment by alexzielenski** [comment]
I don't think so, and I'm not sure that we would pick now to versions earlier than 1.25 when it was introduced.

**Pull request merged**